packages/pwa-buildpack/lib/WebpackTools/plugins/UpwardIncludePlugin.js

```diff
@@ -0,0 +1,145 @@
+const debug = require('../../util/debug').makeFileLogger(__filename);
+const path = require('path');
+const CopyPlugin = require('copy-webpack-plugin');
+const { walkObject } = require('walk-object');
+const { promisify } = require('util');
+const jsYaml = require('js-yaml');
+
+/**
+ * @description webpack plugin that merged UPWARD configurations and
+ * autodetects file assets relied on by those configurations
+ */
+class UpwardIncludePlugin {
+    constructor({ upwardDirs }) {
+        this.upwardDirs = upwardDirs;
+        this.definition = {};
+        debug('created with dirs: %s', upwardDirs);
+    }
+    apply(compiler) {
+        this.compiler = compiler;
+        const { inputFileSystem } = compiler;
+        this.fs = {
+            readFile: promisify(inputFileSystem.readFile.bind(inputFileSystem)),
+            stat: promisify(inputFileSystem.stat.bind(inputFileSystem))
+        };
+        const onRun = () => this.onRun();
+        compiler.hooks.beforeRun.tapPromise('UpwardIncludePlugin', onRun);
+        compiler.hooks.watchRun.tapPromise('UpwardIncludePlugin', onRun);
+    }
+    async onRun() {
+        const { context } = this.compiler.options;
+        this.assetMap = {
+            'upward.yml': {
+                context,
+                from: './upward.yml',
+                to: './upward.yml',
+                transform: () => jsYaml.safeDump(this.definition)
+            }
+        };
+        this.dirs = new Set([...this.upwardDirs, context]);
+        const definitions = await Promise.all(
+            [...this.dirs].map(async dir => {
+                const definition = await this.readUpwardFile(dir);
+                await this.populateAssetMap(dir, definition);
+                return definition;
+            })
+        );
+        Object.assign(this.definition, ...definitions);
+        debug('assigned %s definitions', Object.keys(this.definition));
+
+        debug('assets collection complete, %O', this.assetMap);
+
+        new CopyPlugin(Object.values(this.assetMap), {
+            copyUnmodified: true,
+            logLevel: 'error'
+        }).apply(this.compiler);
+    }
+    extractFileRefs(definition) {
+        const refs = [];
+        walkObject(definition, ({ value, isLeaf }) => {
+            if (isLeaf) {
+                if (typeof value === 'string' && value.startsWith('./')) {
+                    debug('Leaf %s looks like a fs path', value);
+                    refs.push(value);
+                }
+            }
+        });
+        debug('found %s file refs', refs.length);
+        return refs;
+    }
+    async populateAssetMap(dir, definition) {
+        await Promise.all(
+            this.extractFileRefs(definition).map(async ref => {
+                const mapping = await this.validateAndAddRef(dir, ref);
```

Comment:
Likewise, consider `Array.from()` here.

Reply:
I switched to `Array.from` on line 41 based on your other comment, but I'm not sure how I'd do that here.

Reply:
Yeah, it actually wouldn't accomplish anything here. 👍